### PR TITLE
parse new result-set struct

### PR DESCRIPTION
**Pull request by swilly22**
DO NOT MERGE!

### CODE REVIEW

File: src/resultSet.js [hunk 1]
```diff
@@ -1,37 +1,192 @@
...
+        let result_set = raw_result_set[2];
+        let records = new Array(result_set.length);
+        this._results = new Array(result_set.length);
+
+        for(var i = 0; i < result_set.length; i++) {
+        	let row = result_set[i];
+            let record = new Array(row.length);
+            for(var j = 0; j < row.length; j++) {
+            	let cell = row[j];
+                if (this._header[j][0] == ResultSetColumnTypes.COLUMN_SCALAR) {
```
Comment (gkorland) [inline src/resultSet.js]: switch?

File: src/resultSet.js [hunk 2]
```diff
@@ -1,37 +1,192 @@
...
+    }
+
+	parseNode(cell) {
+        // Node ID (integer),
+        // [label string offset (integer)],
+        // [[name, value, value type] X N]
+
+        let node_id = Number(cell[0]);
+        let label = null;
+        if (cell[1].length != 0) {
```
Comment (gkorland) [inline src/resultSet.js]: if (cell[1].length != 0)  ==> if(cell[1].length)

File: src/edge.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
+/**
+ * An edge connecting two nodes.
+ */
+class Edge {
+    constructor(src_node, relation, dest_node, edge_id, properties) {
```
Comment (gkorland) [inline src/edge.js]: Arguments should be camelCase (no underscores)

File: src/edge.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
+/**
+ * An edge connecting two nodes.
+ */
+class Edge {
+    constructor(src_node, relation, dest_node, edge_id, properties) {        
+        this.id = edge_id;
+        this.relation = relation;
+        this.properties = properties;
+        this.src_node = src_node;
```
Comment (gkorland) [inline src/edge.js]: properties should be camelCase

File: src/resultSet.js [hunk 5]
```diff
@@ -1,37 +1,192 @@
...
+                    record[j] = this.parseScalar(cell);
+                }
+                else if (this._header[j][0] == ResultSetColumnTypes.COLUMN_NODE) {
+                    record[j] = this.parseNode(cell);
+                }
+                else if (this._header[j][0] == ResultSetColumnTypes.COLUMN_RELATION) {
+                    record[j] = this.parseEdge(cell);
+                }
+                else {
+                    console.log("Unknown column type.\n");
```
Comment (gkorland) [inline src/resultSet.js]: continue?

File: src/resultSet.js [hunk 6]
```diff
@@ -1,37 +1,192 @@
...
+        let properties = this.parseEntityProperties(cell[4]);
+        return new Edge(src_node_id, relation, dest_node_id, edge_id, properties);
+    }
+
+	parseScalar(cell) {
+        let scalar_type = cell[0];
+        let value = cell[1];
+        let scalar;
+
+        if (scalar_type == ResultSetScalarTypes.PROPERTY_NULL) {
```
Comment (gkorland) [inline src/resultSet.js]: switch?

File: src/resultSet.js [hunk 7]
```diff
@@ -1,37 +1,192 @@
...
+                console.log("Unknown boolean type\n");
+            }
+        }
+
+        else if (scalar_type == ResultSetScalarTypes.PROPERTY_DOUBLE) {
+            scalar = float(value);
+        }
+
+        else if (scalar_type == ResultSetScalarTypes.PROPERTY_UNKNOWN) {
+            console.log("Unknown scalar type\n");
```
Comment (gkorland) [inline src/resultSet.js]: continue?

File: src/resultSet.js [hunk 8]
```diff
@@ -1,37 +1,192 @@
 const Statistics = require("./statistics"),
 	Record = require("./record");
+	Node = require("./node");
+	Edge = require("./edge");
+
+const seasons = {
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/resultSet.js]: oops

File: src/resultSet.js [hunk 9]
```diff
@@ -1,37 +1,192 @@
...
+    SUMMER: 'summer',
+    WINTER: 'winter',
+    SPRING: 'spring',
+    AUTUMN: 'autumn'
+}
+
+const ResultSetColumnTypes = {
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/resultSet.js]: isn't that the way to create an enum in JavaScript?

File: src/resultSet.js (hunk 5 above)
Comment (swilly22) [reply src/resultSet.js]: maybe abort, not sure...

Comment (gkorland) [reply src/resultSet.js]: Only if you throws an Error in this case

File: src/resultSet.js [hunk 10]
```diff
@@ -1,37 +1,185 @@
...
-			this._totalResults = result.length - 1;
-			this._results = new Array(this._totalResults);
-			for (let i = 0; i < this._totalResults; ++i) {
-				this._results[i] = new Record(this._header, result[i + 1]);
-			}
-		}
+        this.parseRecords(resp);
 	}
 
+    parseStringMapping(raw_result_set) {
```
Comment (gkorland) [inline src/resultSet.js]: raw_result_set ==> rawResultSet

File: src/resultSet.js [hunk 11]
```diff
@@ -1,37 +1,185 @@
...
+        // An array of strings, which are referred to
+        // by other parts of the result-set.
+        let string_mapping = raw_result_set[0];
+        return string_mapping;
+	}
+
+	parseHeader(raw_result_set) {
```
Comment (gkorland) [inline src/resultSet.js]: raw_result_set ==> rawResultSet

File: src/resultSet.js [hunk 12]
```diff
@@ -1,37 +1,185 @@
...
+        // An array of column name/column type pairs.
+        let header = raw_result_set[1];
+        return header;
+	}
+
+	parseRecords(raw_result_set) {
```
Comment (gkorland) [inline src/resultSet.js]: raw_result_set ==> rawResultSet

File: src/resultSet.js [hunk 13]
```diff
@@ -1,37 +1,185 @@
...
+            this._statistics = new Statistics(resp[0]);
+        } else{
+    		this.parseResults(resp);
+    		this._resultsCount = this._results.length;
+            this._statistics = new Statistics(resp[3]);
+        }
+	}
+
+	parseResults(resp) {
+        this._string_mapping = this.parseStringMapping(resp);
```
Comment (gkorland) [inline src/resultSet.js]: this._string_mapping ==> this._stringMapping

File: src/resultSet.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (gkorland) [reply src/resultSet.js]: sorry I mixed with Java

File: src/resultSet.js (hunk 7 above)
Comment (swilly22) [reply src/resultSet.js]: Probably should emit an error.

File: src/resultSet.js [hunk 14]
```diff
@@ -1,37 +1,192 @@
...
+
+        return properties;
+    }
+
+	parseNode(cell) {
+        // Node ID (integer),
+        // [label string offset (integer)],
+        // [[name, value, value type] X N]
+
+        let node_id = Number(cell[0]);
```
[the remark this answers is not in the document]
Comment (swilly22) [reply src/resultSet.js]: Isn't cell[0] already a number?

File: src/resultSet.js [hunk 15]
```diff
@@ -1,46 +1,208 @@
...
+        let label = this._graph.getLabel(cell[1][0]);
+        if (label == undefined) {
+            label = await this._graph.fetchAndGetLabel(cell[1][0]);
+        }
+        // let label = null;
```
Comment (swilly22) [inline src/resultSet.js]: Remove comments.

File: src/resultSet.js [hunk 16]
```diff
@@ -1,46 +1,208 @@
...
+    }
+
+    async parseEdge(cell) {
+        // Edge ID (integer),
+        // reltype string offset (integer),
+        // src node ID offset (integer),
+        // dest node ID offset (integer),
+        // [[name, value, value type] X N]
+
+        let edge_id = Number(cell[0]);
```
Comment (swilly22) [inline src/resultSet.js]: Isn't cell[0] already a number?

File: src/resultSet.js [hunk 17]
```diff
@@ -1,46 +1,208 @@
...
+        let relation = this._graph.getRelationship(cell[1]);
+        if (relation == undefined) {
+            relation = await this._graph.fetchAndGetRelationship(cell[1])
+        }
+        let src_node_id = Number(cell[2]);
```
Comment (swilly22) [inline src/resultSet.js]: aren't cell[2] and cell[3] already numbers?

File: src/resultSet.js [hunk 18]
```diff
@@ -1,46 +1,208 @@
...
+        let dest_node_id = Number(cell[3]);
+        let properties = await this.parseEntityProperties(cell[4]);
+        return new Edge(src_node_id, relation, dest_node_id, edge_id, properties);
+    }
+
+    parseScalar(cell) {
+        let scalar_type = cell[0];
+        let value = cell[1];
+        let scalar;
```
Comment (swilly22) [inline src/resultSet.js]: Start with scalar set to undefined

File: src/edge.js [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+class Edge {
+    constructor(srcNode, relation, destNode, edgeId, properties) {        
+        this.id = edgeId;
+        this.relation = relation;
+        this.properties = properties;
+        this.srcNode = srcNode;
+        this.destNode = destNode;
+    }
+
+    toString() {
```
Comment (swilly22) [inline src/edge.js]: If unused, remove.

File: src/node.js [hunk 20]
```diff
@@ -0,0 +1,28 @@
+/**
+ * A node within the garph.
+ */
+class Node {
+    constructor(nodeId, alias, label, properties) {
```
Comment (swilly22) [inline src/node.js]: I think alias is unused, if so please remove.

File: src/node.js [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+        this.id = nodeId;
+        this.alias = alias;
+        this.label = label;
+        this.properties = properties;
+    }
+
+    toString() {
```
Comment (swilly22) [inline src/node.js]: Seems like toString isn't used, if so please remove.

File: .vscode/launch.json [hunk 22]
```diff
@@ -0,0 +1,28 @@
+{
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply .vscode/launch.json]: fixed

File: package.json [hunk 23]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "redisgraph.js",
-	"version": "1.1.4",
+	"version": "1.1.5",
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply package.json]: fixed

File: src/edge.js [hunk 24]
```diff
@@ -0,0 +1,29 @@
+/**
+ * An edge connecting two nodes.
+ */
+class Edge {
+    constructor(srcNode, relation, destNode, edgeId, properties) {
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/edge.js]: fixed

File: src/edge.js (hunk 19 above)
Comment (DvirDukhan) [reply src/edge.js]: removed - using JSON.stringify

File: src/edge.js [hunk 25]
```diff
@@ -0,0 +1,29 @@
...
+        this.srcNode = srcNode;
+        this.destNode = destNode;
+    }
+
+    toString() {
+        let res = "";
+        let props = "";
+
+        for(var i = 0; i < this.properties.length; i++) {
+            for (var entry of map.entries()) {
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/edge.js]: removed - using JSON.stringify

File: src/graph.js [hunk 26]
```diff
@@ -0,0 +1,179 @@
...
+				? host
+				: redis.createClient.apply(redis, [].slice.call(arguments, 1));
+		this._sendCommand = util.promisify(client.send_command).bind(client);
+	}
+
+	/**
+	 * Execute a Cypher query
+	 *
+	 * @param query Cypher query
+	 * @return a result set
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/graph.js]: fixed

File: src/graph.js [hunk 27]
```diff
@@ -0,0 +1,179 @@
...
+
+				return resultSet.parseResponse(res);
+			}
+		);
+	}
+
+	/**
+	 * Deletes the entire graph
+	 *
+	 * @return delete running time statistics
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/graph.js]: fixed

File: src/graph.js [hunk 28]
```diff
@@ -0,0 +1,179 @@
...
+			return resultSet.parseResponse(res);
+		});
+	}
+
+	/**
+	* Calls procedure
+	*
+	* @param procedure Procedure to call
+	* @param args Arguments to pass
+	* @y Yield outputs
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/graph.js]: fixed

File: src/graph.js [hunk 29]
```diff
@@ -0,0 +1,179 @@
...
+	* @return a result set
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/graph.js]: fixed

File: .gitignore [hunk 30]
```diff
@@ -1,7 +1,9 @@
 /.settings/
 /node_modules/
```
Comment (swilly22) [inline .gitignore]: Remove '/' for all folders.

File: src/edge.js [hunk 31]
```diff
@@ -0,0 +1,21 @@
+/**
+ * An edge connecting two nodes.
+ */
+class Edge {
+    constructor(srcNode, relation, destNode, properties) {
+        this.id = "";
```
Comment (swilly22) [inline src/edge.js]: I think setting id to undefined is better than an empty string.

File: src/graph.js [hunk 32]
```diff
@@ -0,0 +1,176 @@
...
+
+	/**
+	 * auxilary function to extract string(s) data from procedures such as:
+	 * db.labels, db.propertyKeys and db.relationshipTypes
+	 * @param  resultSet - a procedure result set
+	 */
+	_extractStrings(resultSet) {
+		var strings = [];
+		while (resultSet.hasNext()) {
+
```
Comment (swilly22) [inline src/graph.js]: Remove empty line

File: src/graph.js [hunk 33]
```diff
@@ -0,0 +1,176 @@
...
+			var str = resultSet.next().get(0);
```
Comment (swilly22) [inline src/graph.js]: strings.push(resultSet.next().getString(0));

File: src/graph.js [hunk 34]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	/**
+	 * Execute a Cypher query (async)
+	 *
+	 * @param query Cypher query
+	 * @return a promise contains a result set
+	 */
+	async query(query) {
+		var res = await this._sendCommand("graph.QUERY", [this._graphId, query, "--compact"]);
```
Comment (swilly22) [inline src/graph.js]: Nice!

File: src/graph.js [hunk 35]
```diff
@@ -0,0 +1,176 @@
...
+		return resultSet.parseResponse(res);
+	}
+
+	/**
+	 * Deletes the entire graph (async)
+	 *
+	 * @return a promise contains the delete operation running time statistics
+	 */
+	async deleteGraph() {
+		var res = await this._sendCommand("graph.DELETE", [this._graphId]);
```
Comment (swilly22) [inline src/graph.js]: I don't think graph.DELETE returns a result-set.

File: src/graph.js [hunk 36]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	/**
+	 * Retrieves all labels in graph.
+	 *
+	 * @return array of labels.
+	 */
+	async labels() {
+		var response = await this.callProcedure("db.labels");
+		this._labels = this._extractStrings(response);
```
Comment (swilly22) [inline src/graph.js]: This functions doesn't return anything contradicting the comment.

File: src/graph.js [hunk 37]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	/**
+	 * Retrieves all relationship types in graph.
+	 *
+	 * @return array of relationship types.
+	 */
+	async relationshipTypes() {
+		var response = await this.callProcedure("db.relationshipTypes")
+		this._relationshipTypes = this._extractStrings(response);
```
Comment (swilly22) [inline src/graph.js]: Nothing is returned...

File: src/graph.js [hunk 38]
```diff
@@ -0,0 +1,176 @@
...
+	}
+
+	/**
+	 * Retrieves all properties in graph.
+	 *
+	 * @return array of properties.
+	 */
+	async propertyKeys() {
+		var response = await this.callProcedure("db.propertyKeys");
+		this._properties = this._extractStrings(response);
```
Comment (swilly22) [inline src/graph.js]: Nothing is returned, see function comment.

File: src/node.js [hunk 39]
```diff
@@ -0,0 +1,20 @@
+/**
+ * A node within the garph.
+ */
+class Node {
+    constructor(label, properties) {
+        this.id = "";
```
Comment (swilly22) [inline src/node.js]: Set id to undefined.

File: src/resultSet.js [hunk 40]
```diff
@@ -1,46 +1,219 @@
...
-	next() {
-		return this._results[this._position++];
-	}
-
-	getStatistics() {
-		return this._statistics;
-	}
+    constructor(graph) {
+        //_graph is graph api
+        this._graph = graph;
```
Comment (swilly22) [inline src/resultSet.js]: Comment at the same line as the prop,
this._graph = graph;    //_graph is graph api

File: src/graph.js [hunk 41]
```diff
@@ -21,6 +21,10 @@ class Graph {
 		this._relationshipTypes = []; 	// List of relation types.
 		this._properties = [];        	// List of properties.
 
+		this._labelsPromise = undefined;
```
Comment (swilly22) [inline src/graph.js]: Add comments for _labelsPromise, _propertyPromise and _relationshipPromise explaining what are they used for,
similar to any other class member.

File: src/graph.js [hunk 42]
```diff
@@ -60,6 +64,9 @@ class Graph {
 	 */
 	async deleteGraph() {
 		var res = await this._sendCommand("graph.DELETE", [this._graphId]);
+		this._labels = [];
```
Comment (swilly22) [inline src/graph.js]: Simple comment: "clear internal graph state"

File: src/graph.js [hunk 43]
```diff
@@ -81,24 +88,58 @@ class Graph {
 	 * Retrieves all labels in graph.
 	 */
 	async labels() {
-		var response = await this.callProcedure("db.labels");
-		this._labels = this._extractStrings(response);
+		if (this._labelsPromise == undefined) {
+			console.info("fetching node labels")
```
Comment (swilly22) [inline src/graph.js]: Remove this print

File: src/graph.js [hunk 44]
```diff
@@ -81,24 +88,58 @@ class Graph {
...
+			this._labelsPromise = this.callProcedure("db.labels").then(response => {
+				return this._extractStrings(response);
+			})
+			this._labels = await (this._labelsPromise);
+			this._labelsPromise = undefined;
+		}
+		else {
+			console.info("waiting for label promise");
```
Comment (swilly22) [inline src/graph.js]: Remove print

File: src/graph.js [hunk 45]
```diff
@@ -81,24 +88,58 @@ class Graph {
...
+		if (this._relationshipPromise == undefined) {
+			console.info("fetching relationship types");
+			this._relationshipPromise = this.callProcedure("db.relationshipTypes").then(response => {
+				return this._extractStrings(response);
+			});
+			this._relationshipTypes = await (this._relationshipPromise);
+			this._relationshipPromise = undefined;
+		}
+		else {
+			console.info("waiting for relationship promise");
```
Comment (swilly22) [inline src/graph.js]: Remove print

File: src/graph.js [hunk 46]
```diff
@@ -81,24 +88,58 @@ class Graph {
...
 	}
 
 	/**
 	 * Retrieves all properties in graph.
 	 */
 	async propertyKeys() {
-		var response = await this.callProcedure("db.propertyKeys");
-		this._properties = this._extractStrings(response);
+		if (this._propertyPromise == undefined) {
+			console.info("fetching property names");
```
Comment (swilly22) [inline src/graph.js]: Remove print

File: src/graph.js [hunk 47]
```diff
@@ -81,24 +88,58 @@ class Graph {
...
+			this._propertyPromise = this.callProcedure("db.propertyKeys").then(response => {
+				return this._extractStrings(response);
+			})
+			this._properties = await this._propertyPromise;
+			this._propertyPromise = undefined;
+		}
+		else{
+			console.info("waiting for property promise");
```
Comment (swilly22) [inline src/graph.js]: Remove print

File: src/resultSet.js [hunk 48]
```diff
@@ -114,8 +114,7 @@ class ResultSet {
             let prop = props[i];
             var propIndex = prop[0];
             let prop_name = this._graph.getProperty(propIndex);
-            if (prop_name == undefined) {
-                console.error("fetching property labels");
+            while (prop_name == undefined) {
```
Comment (swilly22) [inline src/resultSet.js]: Consider switching from while to a for loop going from 1 to X, after X tries, if we didn't managed to get the value, log and continue with an undefined value, apply this to all 3 fetches.

**Pull request merged**